Document AOT analysis warnings (#30967)
dotnet/docs: docs/core/deploying/native-aot/warnings/il3054.md

---
title: "IL3054: Generic expansion to a method or type was aborted due to generic recursion"

> description: "Infinite generic recursion is not supported when publishing as native AOT"

Reviewer: ```suggestion
description: "Learn about warning IL3054: Infinite generic recursion is not supported when publishing as native AOT."
```

ms.date: 09/01/2022
ms.topic: reference
ms.author: michals
author: MichalStrehovsky
f1_keywords:
  - "IL3054"
---
# IL3054: Generic expansion to a method or type was aborted due to generic recursion

## Cause

Methods on generic types and generic methods that are instantiated over different types are supported by different native code bodies specialized for the given type parameter.

It is possible to form a cycle between generic instantiations in a way that the number of native code bodies required to support the application becomes unbounded. Because native AOT deployments require generating all native method bodies at the time of publishing the application, this would require compiling an infinite number of methods.

> When the AOT compilation process detects such unbounded growth, it cuts off the growth by generating a throwing method. If the application at runtime goes beyond the cutoff point, an exception will be thrown.

Reviewer: ```suggestion
When the AOT compilation process detects such unbounded growth, it cuts off the growth by generating a throwing method. If the application goes beyond the cutoff point at run time, an exception is thrown.
```

> Even though it's unlikely the throwing method body would be reached at runtime, it is advisable to remove the generic recursion by restructuring the code. Generic recursion negatively affects compilation speed and the size of the output executable.

Reviewer: ```suggestion
Even though it's unlikely the throwing method body will be reached at run time, it's advisable to remove the generic recursion by restructuring the code. Generic recursion negatively affects compilation speed and the size of the output executable.
```

## Example

> Following program will work correctly for input "2", but throw an exception for input "100".

Reviewer: ```suggestion
The following program will work correctly for input "2" but throws an exception for input "100".
```

```csharp
// AOT analysis warning IL3054:
// Program.<<Main>$>g__CauseGenericRecursion|0_0<Struct`1<Struct`1<Struct`1<Struct`1<Int32>>>>>(Int32):
// Generic expansion to 'Program.<<Main>$>g__CauseGenericRecursion|0_0<Struct`1<Struct`1<Struct`1<Struct`1<Struct`1<Int32>>>>>>(Int32)'
// was aborted due to generic recursion. An exception will be thrown at runtime if this codepath
// is ever reached. Generic recursion also negatively affects compilation speed and the size of
// the compilation output. It is advisable to remove the source of the generic recursion
// by restructuring the program around the source of recursion. The source of generic recursion
// might include: 'Program.<<Main>$>g__CauseGenericRecursion|0_0<T>(Int32)

using System;

int number = int.Parse(Console.ReadLine());
Console.WriteLine(CauseGenericRecursion<int>(number));

static int CauseGenericRecursion<T>(int i) => 1 + CauseGenericRecursion<Struct<T>>(i - 1);

struct Struct<T> { }
```

> The behavior of the program at runtime for input "100":

Reviewer: ```suggestion
The behavior of the program at run time for input "100":
```
